### PR TITLE
Clean cabal project

### PR DESCRIPTION
**Pull request by jneira**
* Clean main cabal.project, removing the `shake-bench` submodule and *all* the `allow-newer`
* Create a specific project for run the benchmarks with all the workaround needed for build it:
  * All the allow-newer
  * And the submodule `shake-bench`, which now is being built, with all dependencies  in all builds (with or without benchs)

The idea is backport thia to 1.3.0-hackage to help the hackage release (#1990)

### CODE REVIEW

File: cabal-benchmark.project [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    diagrams-core:base,
+    diagrams-core:lens,
+    diagrams-lib:base,
+    diagrams-lib:lens,
+    diagrams-postscript:base,
+    diagrams-postscript:lens,
+    diagrams-svg:base,
+    diagrams-svg:lens,
+    dual-tree:base,
+    svg-builder:base
```
Comment (pepeiborra) [inline cabal-benchmark.project:59]: Are these still needed though? The upstream issue was fixed recently: https://github.com/diagrams/diagrams-core/issues/112

**Review by pepeiborra — COMMENTED** [review]
How is `shake-bench` blocking the Hackage release? Please explain the issue and I will try to find a solution that doesn't introduce a new `cabal-foo.project` file.

**Comment by jneira** [comment]
> 
> 
> How is `shake-bench` blocking the Hackage release? Please explain the issue and let's try to find a solution that doesn't introduce a new `cabal-foo.project` file.

It is not blocking the release directly but disturbing it with bench only releated allow-newer and a submodule, which is being taken in account for *all* builds even if they are not running or building benchmarks.
Otoh that makes you cant run the benchmarks against the hackage released version of ghcide.
I can restore all the workarounds needed to run the benchmarks in the main cabal.project file (with comments of course, to no be confused with other reason to add allow-newer), but i thought separate it would be a good motivation to try to remove the workarounds, which are since long time ago and seems to stay with us forever.

File: cabal-benchmark.project (hunk 1 above)
Comment (jneira) [reply cabal-benchmark.project:59]: will try it

Comment (jneira) [reply cabal-benchmark.project:59]: without the allow-newer for diagrams-core i got

```
PS D:\hls> cabal bench --project-file .\cabal-benchmark.project
Resolving dependencies...
cabal-3.4.0.0.exe: Could not resolve dependencies:
[__0] trying: ghc-api-compat-8.6.1 (user goal)
[__1] trying: base-4.14.2.0/installed-4.14.2.0 (dependency of ghc-api-compat)
[__2] trying: shake-bench-0.1.0.0 (user goal)
[__3] trying: diagrams-svg-1.4.3 (dependency of shake-bench)
[__4] trying: diagrams-core-1.5.0 (dependency of diagrams-svg)
[__5] next goal: Chart-diagrams (dependency of shake-bench)
[__5] rejecting: Chart-diagrams-1.9.3 (conflict: diagrams-core==1.5.0,
Chart-diagrams => diagrams-core>=1.3 && <1.5)
[__5] skipping: Chart-diagrams-1.9.2, Chart-diagrams-1.9.1,
Chart-diagrams-1.9, Chart-diagrams-1.8.3, Chart-diagrams-1.8.2,
Chart-diagrams-1.8.1, Chart-diagrams-1.8, Chart-diagrams-1.7.1,
Chart-diagrams-1.7, Chart-diagrams-1.6, Chart-diagrams-1.5.4,
Chart-diagrams-1.5.1, Chart-diagrams-1.5, Chart-diagrams-1.4,
Chart-diagrams-1.3.3, Chart-diagrams-1.3.2, Chart-diagrams-1.3.1,
Chart-diagrams-1.3, Chart-diagrams-1.2.4, Chart-diagrams-1.2.3,
Chart-diagrams-1.2.2, Chart-diagrams-1.2, Chart-diagrams-1.1,
Chart-diagrams-1.0 (has the same characteristics that caused the previous
version to fail: excludes 'diagrams-core' version 1.5.0)
[__5] fail (backjumping, conflict set: Chart-diagrams, diagrams-core,
shake-bench)
After searching the rest of the dependency tree exhaustively, these were the
goals I've had most trouble fulfilling: diagrams-core, base, Chart-diagrams,
shake-bench, diagrams-svg, ghc-api-compat
Try running with --minimize-conflict-set to improve the error message.
```

After trying several combinations i only get it to build enabling all of them maybe a more detailed analysis could make let us know the way to remove them.

Comment (jneira) [reply cabal-benchmark.project]: I've got to remove some other entries from `allow-newer`

**Comment by jneira** [comment]
@pepeiborra ok, i've unified both `cabal*.project` again. Only would like to have some issue or plan to eventually remove the `allow-newer` i've not removed in this pr.

**Comment by jneira** [comment]
@Mergifyio backport 1.3.0-hackage

**Comment by pepeiborra** [comment]
The allow-newer for all the diagrams ecosystem should not be needed anymore, did you check?

**Comment by pepeiborra** [comment]
> > 
> 
> > 
> 
> > How is `shake-bench` blocking the Hackage release? Please explain the issue and let's try to find a solution that doesn't introduce a new `cabal-foo.project` file.
> 
> 
> 
> It is not blocking the release directly but disturbing it with bench only releated allow-newer and a submodule, which is being taken in account for *all* builds even if they are not running or building benchmarks.

The allow-newer might be superfluous, let's find out. 

To avoid building shake-bench, just be more precise and list the targets that you wish to build: "cabal build haskell-language-server"

> Otoh that makes you cant run the benchmarks against the hackage released version of ghcide.

Not sure what you mean. Care to elaborate,

> I can restore all the workarounds needed to run the benchmarks in the main cabal.project file (with comments of course, to no be confused with other reason to add allow-newer), but i thought separate it would be a good motivation to try to remove the workarounds, which are since long time ago and seems to stay with us forever.
> 

What workarounds are you referring to?

**Comment by jneira** [comment]
> The allow-newer for all the diagrams ecosystem should not be needed anymore, did you check?

yeah, see https://github.com/haskell/haskell-language-server/pull/2097#discussion_r689383367

**Comment by jneira** [comment]
> To avoid building shake-bench, just be more precise and list the targets that you wish to build: \"cabal build haskell-language-server\

in my tests, no matter if you set a specific target or use `--disable-benchmarks` shake-bench is taken in account to compute the build plan, so it makes all builds fail if cabal does not find a plan including it

**Comment by jneira** [comment]
> Not sure what you mean. Care to elaborate,

due to allow-never you can't run the benchmarks downloading the package (s) from hackage

**Comment by jneira** [comment]
> What workarounds are you referring to?

for me allow-newer and remote source repo packages in cabal.project are workarounds to be removed asap

**Comment by pepeiborra** [comment]
> > To avoid building shake-bench, just be more precise and list the targets that you wish to build: "cabal build haskell-language-server\
> 
> in my tests, no matter if you set a specific target or use `--disable-benchmarks` shake-bench is taken in account to compute the build plan, so it makes all builds fail if cabal does not find a plan including it

Understood. I have opened #2101 to remove as many as possible, there are still 2 left.

**Review by pepeiborra — APPROVED** [review]

**Pull request merged**